src/Illuminate/Events/Dispatcher.php

```diff
@@ -87,19 +87,55 @@ public function __construct(ContainerContract $container = null)
     public function listen($events, $listener = null)
     {
         if ($events instanceof Closure) {
-            return collect($this->firstClosureParameterTypes($events))
-                ->each(function ($event) use ($events) {
-                    $this->listen($event, $events);
-                });
-        } elseif ($events instanceof QueuedClosure) {
-            return collect($this->firstClosureParameterTypes($events->closure))
-                ->each(function ($event) use ($events) {
-                    $this->listen($event, $events->resolve());
-                });
-        } elseif ($listener instanceof QueuedClosure) {
+            return $this->handleClosureEvents($events);
+        }
+
+        if ($events instanceof QueuedClosure) {
+            return $this->handleQueuedClosureEvents($events);
+        }
+
+        if ($listener instanceof QueuedClosure) {
             $listener = $listener->resolve();
         }
 
+        return $this->handleEventArray($events, $listener);
+    }
+
+    /**
+     * Handle closure events.
+     *
+     * @param  \Closure|string|array  $events
+     */
+    private function handleClosureEvents($events)
```

Comment:
If this abstraction if truly going to be used, I think that this method would be `protected` instead of `private`, and since we have already type checked `$events` in `listen()`, I would think that the Closure type should be enforced: `handleClosureEvents(Closure $events)`

This would also apply to `handleQueuedClosureEvents` with `QueuedClosure` (and I would keep the variable name `$events` instead of changing it to `$queuedClosureEvents`).